includes/utils/class-amp-dom-utils.php

```diff
@@ -141,9 +158,62 @@ function( $noscript_matches ) {
 			}
 		}
 
+		// Make sure there is a head and a body.
+		$head = $dom->getElementsByTagName( 'head' )->item( 0 );
+		if ( ! $head ) {
+			$head = $dom->createElement( 'head' );
+			$dom->documentElement->insertBefore( $head, $dom->documentElement->firstChild );
+		}
+		$body = $dom->getElementsByTagName( 'body' )->item( 0 );
+		if ( ! $body ) {
+			$body = $dom->createElement( 'body' );
+			$dom->documentElement->appendChild( $body );
+		}
+		self::move_invalid_head_nodes_to_body( $head, $body );
+
 		return $dom;
 	}
 
+	/**
+	 * Move elements not allowed in the head to the body.
+	 *
+	 * Apparently PHP's DOM is more lenient when parsing HTML to allow nodes in the HEAD which do not belong. A proper
+	 * HTML5 parser should rather prematurely short-circuit the HEAD when it finds an illegal element.
+	 *
+	 * @param DOMElement $head HEAD element.
+	 * @param DOMElement $body BODY element.
+	 */
+	private static function move_invalid_head_nodes_to_body( $head, $body ) {
+		// Walking backwards is to make it easier to prepend them to the body in the expected order.
```

Comment:
Consider: "Walking backwards makes it easier to move elements in the expected order"